### PR TITLE
Trigger Missing Jobs Configuration and refactoring for pulse_listener.

### PR DESCRIPTION
**Pull request by vaibhavmagarwal**
Not ready for review. Will depend on fill-in treeherder patch that is yet to be checked-in.

### CODE REVIEW

**Comment by vaibhavmagarwal** [comment]
r? @adusca @chmanchester

Do not merge this in. We need to do a new mozci release before merging this in.

**Comment by armenzg** [comment]
Hi @vaibhavmagarwal what are the next steps on this PR?
Thanks!

**Comment by vaibhavmagarwal** [comment]
@armenzg I have updated the mozci version, and this PR is now ready for review.
r?

**Comment by armenzg** [comment]
Hi @vaibhavmagarwal after looking at the patch I believe it is better for @adusca to review it.
I don't know pulse_actions enough to give you good feedback.
From a quick look it looks like it makes sense.

**Comment by armenzg** [comment]
fyi I could review the patch but I would need to spend more time reading pulse_actions code and it might take me a decent amount of time to grasp the design model which can take a bit.

**Comment by vaibhavmagarwal** [comment]
@adusca could you review this?

File: pulse_actions/resultset_action.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from pulse_actions.worker import run_exchange_topic
```
Comment (vaibhavmagarwal) [inline pulse_actions/resultset_action.py]: This file is to be called:
    python pulse_actions/resultset_action.py

for listening to pulse messages on exchange/treeherder/v1/resultset-actions exchange.

**Comment by vaibhavmagarwal** [comment]
This PR handles just no. 1 part of approach in https://github.com/adusca/pulse_actions/issues/6#issuecomment-124241050

**Comment by vaibhavmagarwal** [comment]
It will be good to have no.1 part of approach running atleast for time being, i.e.,
*\* triggering missing downstream jobs for which upstream jobs exists, 
*\* triggering missing upstream jobs
since it solves the problem partially, and we will reduce atleast some pain for the sheriffs.

File: setup.py [hunk 2]
```diff
@@ -26,6 +27,7 @@
       url='https://github.com/adusca/pulse_actions',
       entry_points={
           'console_scripts': [
-              'run-pulse-actions = pulse_actions.worker:main'
+              'run-job-actions = pulse_actions.worker:main',
+              'run-resultset-actions = pulse_actions.resultset_action:main'
               ],
```
Comment (vaibhavmagarwal) [inline setup.py:31]: @adusca I have added a console script now.

**Comment by vaibhavmagarwal** [comment]
@adusca could you give feedback on this patch? I am not sure how to include "multiple exchanges", how should I go about doing that? Thanks!
FYI, this patch runs properly, if I do:
python pulse_actions/worker.py (for buildbot messages)
python pulse_actions/resultset_action.py (for resultset messages)

File: pulse_actions/worker.py [hunk 3]
```diff
@@ -55,33 +55,39 @@ def handler_with_dry_run(data, message):
...
-        topic=options['topic'],
+        exchange=options[topic_base]['exchange'],
+        topic=options[topic_base]['topic'],
         event_handler=handler_function,
+        topic_base=topic_base,
         dry_run=True)
 
+def main():
+    run_exchange_topic(topic_base='buildbot')
+
```
Comment (adusca) [inline pulse_actions/worker.py:106]: Can you make topic_base a command-line argument that defaults to 'buildbot'?

That way we don't need pulse_actions/resultset_action.py , just a different command-line argument.

File: pulse_actions/handlers/treeherder_resultset.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+from mozci.mozci import trigger_missing_jobs_for_revision
+from thclient import TreeherderClient
+from pulse_actions.publisher import MessageHandler
+
+
```
Comment (adusca) [inline pulse_actions/handlers/treeherder_resultset.py:5]: I think we need to set MEMORY_SAVING_MODE to True somewhere so it can run on Heroku.

**Comment by adusca** [comment]
My mozillapulse patch for multiple exchanges landed, now we just have to wait for a new mozillapulse release.

After that, to have one listener for multiple exchanges, all you have to do is:

exchange: [list of exchange]
topic: [corresponding list of topics] 

in runtime_config.json.

Then we will need a handler function that calls the appropriate handler function per message (if it is a exchange/build/normalized message, call the exchange/build/normalized function, etc).

**Comment by adusca** [comment]
r+ after the small changes. We need to figure out if we are going to have 2 workers on Heroku or if we are going to use multiple exchanges (and wait for the next mozillapulse release).

**Comment by adusca** [comment]
We now can use MozillaPulse 1.1 to have a consumer for multiple exchanges.

**Comment by vaibhavmagarwal** [comment]
@adusca, I have made the changes.

**Comment by chmanchester** [comment]
Let's have a separate worker for trigger_missing_jobs for now -- we'll need the multiple exchange functionality within this worker later when we implement job tracking. This will at least give us the ability to filter logs based on the originating worker when we're debugging one feature or the other.

**Comment by vaibhavmagarwal** [comment]
@adusca can we merge this patch in and test it out? Please let me know what I need to do for its deployment.

**Comment by chmanchester** [comment]
I provisioned an extra dyno for pulse-actions.

**Pull request merged**